File: CK3ToEU4/Source/Configuration/Configuration.cs
using System;
using System.IO;
using commonItems;
using CommonFunctions = commonItems.CommonFunctions;

namespace CK3ToEU4.NET.Source.Configuration;

public enum StartDate
{
    EU = 1,
    CK = 2,
    Manual = 3,
}

public enum SplitVassals
{
    Yes = 1,
    No = 2
}
public enum DeJure
{
    Enabled = 1,
    Disabled = 2
}
public enum Development
{
    Import = 1,
    Vanilla = 2
}
public enum MultiProvDevTransfer
{
    Default = 1,
    MP = 2
}
public enum Sunset
{
    Disabled = 1,
    Active = 2
}
public enum IAmHRE
{
    HRE = 1,
    Byzantium = 2,
    Rome = 3,
    Custom = 4,
    None = 5
}
public enum ShatterEmpires
{
    None = 1,
    All = 2,
    Custom = 3
}
public enum ShatterLevel
{
    Duchy = 1,
    Kingdom = 2
}
public enum ShatterHRELevel
{
    Duchy = 1,
    Kingdom = 2
}
public enum Siberia
{
    SmallSiberia = 1,
    ClearSiberia = 2,
    LeaveSiberia = 3
}
public enum Institutions
{
    Historic = 1,
    Dynamic = 2
}
public enum DynasticNames
{
    Enabled = 1,
    Disabled = 2
}
public enum DiscoveredBy
{
    Vanilla = 1,
    Dynamic = 2
}

public class Config
{
    public string SaveGamePath { get; private set; } = string.Empty;
    public string CK3Path { get; private set; } = string.Empty;
    public string CK3DocPath { get; private set; } = string.Empty;
    public string EU4Path { get; private set; } = string.Empty;
    public string OutputName { get; private set; } = string.Empty;
    public StartDate StartDateOption { get; private set; } = StartDate.EU;
    public IAmHRE HRE { get; private set; } = IAmHRE.HRE;
    public ShatterEmpires ShatterEmpires { get; private set; } = ShatterEmpires.None;
    public ShatterHRELevel ShatterHRELevel { get; private set; } = ShatterHRELevel.Duchy;
    public ShatterLevel ShatterLevel { get; private set; } = ShatterLevel.Duchy;
    public Siberia Siberia { get; private set; } = Siberia.ClearSiberia;
    public Sunset Sunset { get; private set; } = Sunset.Disabled;
    public Institutions DynamicInstitutions { get; private set; } = Institutions.Historic;
    public Development Development { get; private set; } = Development.Import;
    public MultiProvDevTransfer MultiProvinceDevelopmentTransfer { get; private set; } = MultiProvDevTransfer.Default;
    public DeJure DeJure { get; private set; } = DeJure.Enabled;
    public SplitVassals SplitVassals { get; private set; } = SplitVassals.Yes;
    public DynasticNames DynasticNames { get; private set; } = DynasticNames.Enabled;
    public DiscoveredBy DiscoveredBy { get; private set; } = DiscoveredBy.Vanilla;
    
    public Config(ConverterVersion converterVersion)
    {
        Logger.Info("Reading configuration file");
        var parser = new Parser();
        RegisterKeys(parser);
        parser.ParseFile("configuration.txt");

        SetOutputName();
        VerifyCK3Path();
        VerifyCK3Version(converterVersion);
        VerifyEU4Path();
        VerifyEU4Version(converterVersion);
        Logger.Progress(3);
    }

    public Config(BufferedReader reader)
    {
        var parser = new Parser();
        RegisterKeys(parser);
        parser.ParseStream(reader);

        SetOutputName();
    }

    private void RegisterKeys(Parser parser)
    {
        parser.RegisterKeyword("SaveGame", reader =>
        {
            SaveGamePath = reader.GetString();
            Logger.Info($"Save Game set to: {SaveGamePath}");
        });
        parser.RegisterKeyword("CK3directory", reader => CK3Path = reader.GetString());
        parser.RegisterKeyword("EU4directory", reader => EU4Path = reader.GetString());
        parser.RegisterKeyword("CK3DocDirectory", reader => CK3DocPath = reader.GetString());
        parser.RegisterKeyword("output_name", reader => {
            OutputName = reader.GetString();
            Logger.Info($"Output name set to: {OutputName}");
        });
        parser.RegisterKeyword("start_date", reader =>
        {
            var startDateString = reader.GetString();
            StartDateOption = (StartDate)int.Parse(startDateString);
            Logger.Info($"Start date set to: {startDateString}");
        });
        parser.RegisterKeyword("i_am_hre", reader => {
            var hreString = reader.GetString();
            HRE = (IAmHRE)int.Parse(hreString);
            Logger.Info($"HRE set to: {hreString}");
        });
        parser.RegisterKeyword("dejure", reader => {
            var deJureString = reader.GetString();
            DeJure = (DeJure)int.Parse(deJureString);
            Logger.Info($"DeJure set to: {deJureString}");
        });
        parser.RegisterKeyword("split_vassals", reader => {
            var splitVassalsString = reader.GetString();
            SplitVassals = (SplitVassals)int.Parse(splitVassalsString);
            Logger.Info($"Split Vassals set to: {splitVassalsString}");
        });
        parser.RegisterKeyword("shatter_empires", reader => {
            var shatterEmpiresString = reader.GetString();
            ShatterEmpires = (ShatterEmpires)int.Parse(shatterEmpiresString);
            Logger.Info($"Shatter Empires set to: {shatterEmpiresString}");
        });
        parser.RegisterKeyword("shatter_level", reader => {
            var shatterLevelString = reader.GetString();
            ShatterLevel = (ShatterLevel)int.Parse(shatterLevelString);
            Logger.Info($"Shatter Level set to: {shatterLevelString}");
        });
        parser.RegisterKeyword("shatter_hre_level", reader => {
            var shatterHRELevelString = reader.GetString();
            ShatterHRELevel = (ShatterHRELevel)int.Parse(shatterHRELevelString);
            Logger.Info($"Shatter HRE Level set to: {shatterHRELevelString}");
        });
        parser.RegisterKeyword("siberia", reader => {
            var siberiaString = reader.GetString();
            Siberia = (Siberia)int.Parse(siberiaString);
            Logger.Info($"Siberia set to: {siberiaString}");
        });
        parser.RegisterKeyword("sunset", reader => {
            var sunsetString = reader.GetString();
            Sunset = (Sunset)int.Parse(sunsetString);
            Logger.Info($"Sunset set to: {sunsetString}");
        });
        parser.RegisterKeyword("dynamicInstitutions", reader => {
            var dynamicInstitutionsString = reader.GetString();
            DynamicInstitutions = (Institutions)int.Parse(dynamicInstitutionsString);
            Logger.Info($"Dynamic Institutions set to: {dynamicInstitutionsString}");
        });
        parser.RegisterKeyword("development", reader => {
            var developmentString = reader.GetString();
            Development = (Development)int.Parse(developmentString);
            Logger.Info($"Development set to: {developmentString}");
        });
        parser.RegisterKeyword("multiprovdevtransfer", reader => {
            var multiProvDevTransferString = reader.GetString();
            MultiProvinceDevelopmentTransfer = (MultiProvDevTransfer)int.Parse(multiProvDevTransferString);
            Logger.Info($"Multiple Province Dev Transfer set to: {multiProvDevTransferString}");
        });
        parser.RegisterKeyword("dynasticNames", reader => {
            var dynasticNamesString = reader.GetString();
            DynasticNames = (DynasticNames)int.Parse(dynasticNamesString);
            Logger.Info($"Dynastic Names set to: {dynasticNamesString}");
        });
        parser.RegisterKeyword("discoveredBy", reader => {
            var discoveredByString = reader.GetString();
            DiscoveredBy = (DiscoveredBy)int.Parse(discoveredByString);
            Logger.Info($"Discovered By set to: {discoveredByString}");
        });
        parser.RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
    }

    private void VerifyCK3Path()
    {
        if (!Directory.Exists(CK3Path))
            throw new Exception(CK3Path + " does not exist!");
        // TODO: OSX and Linux paths are speculative
        if (!File.Exists(CK3Path + "/binaries/ck3.exe") &&
            !File.Exists(CK3Path + "/CK3game") &&
            !File.Exists(CK3Path + "/binaries/ck3"))
            throw new Exception(CK3Path + " does not contain Crusader Kings 3!");
        if (!File.Exists(CK3Path + "/game/map_data/positions.txt"))
            throw new Exception(CK3Path + " does not appear to be a valid CK3 install!");
        Logger.Info($"\tCK3 install path is {CK3Path}");
        CK3Path += "/game/"; // We're adding "/game/" since all we ever need from now on is in that subdirectory.
    }

    private void VerifyEU4Path()
    {
        if (!Directory.Exists(EU4Path))
            throw new Exception(EU4Path + " does not exist!");
        if (!File.Exists(EU4Path + "/eu4.exe") && !File.Exists(EU4Path + "/eu4"))
            throw new Exception(EU4Path + " does not contain Europa Universalis 4!");
        if (!File.Exists(EU4Path + "/map/positions.txt"))
            throw new Exception(EU4Path + " does not appear to be a valid EU4 install!");
        Logger.Info($"\tEU4 install path is {EU4Path}");
    }

    private void SetOutputName()
    {
        if (string.IsNullOrEmpty(OutputName))
        {
            OutputName = CommonFunctions.TrimPath(SaveGamePath);
        }

        OutputName = CommonFunctions.TrimExtension(OutputName);
        OutputName = CommonFunctions.ReplaceCharacter(OutputName, '-');
        OutputName = CommonFunctions.ReplaceCharacter(OutputName, ' ');

        OutputName = CommonFunctions.NormalizeUTF8Path(OutputName);
        Logger.Info($"Using output name {OutputName}");
    }

    private void VerifyCK3Version(ConverterVersion converterVersion)
    {
        var ck3Version = GameVersion.ExtractVersionFromLauncher(CK3Path + "../launcher/launcher-settings.json");
        if (ck3Version is null)
        {
            Logger.Error("CK3 version could not be determined, proceeding blind!");
            return;
        }

        Logger.Info($"CK3 version: {ck3Version.ToShortString()}");

        if (converterVersion.MinSource > ck3Version)
        {
            Logger.Error($"CK3 version is v{ck3Version.ToShortString()}" +
                ", converter requires minimum v" +
                $"{converterVersion.MinSource.ToShortString()}!");
            throw new Exception("Converter vs CK3 installation mismatch!");
        }

        if (!converterVersion.MaxSource.IsLargerishThan(ck3Version))
        {
            Logger.Error($"CK3 version is v{ck3Version.ToShortString()}" +
                ", converter requires maximum v" +
                $"{converterVersion.MaxSource.ToShortString()}!");
            throw new Exception("Converter vs CK3 installation mismatch!");
        }
    }

    private void VerifyEU4Version(ConverterVersion converterVersion)
    {
        var eu4Version = GameVersion.ExtractVersionFromLauncher(EU4Path + "/launcher-settings.json");
        if (eu4Version is null)
        {
            Logger.Error("EU4 version could not be determined, proceeding blind!");
            return;
        }

        Logger.Info($"EU4 version: {eu4Version.ToShortString()}");

        if (converterVersion.MinTarget > eu4Version)
        {
            Logger.Error($"EU4 version is v{eu4Version.ToShortString()}" +
                ", converter requires minimum v" +
                $"{converterVersion.MinTarget.ToShortString()}!");
            throw new Exception("Converter vs EU4 installation mismatch!");
        }

        if (!converterVersion.MaxTarget.IsLargerishThan(eu4Version))
        {
            Logger.Error($"EU4 version is v{eu4Version.ToShortString()}" +
                ", converter requires maximum v" +
                $"{converterVersion.MaxTarget.ToShortString()}!");
            throw new Exception("Converter vs EU4 installation mismatch!");
        }
    }
}

File: CK3ToEU4/Source/Configuration/Configuration.cpp
#include "Configuration.h"
#include "Color.h"
#include "CommonFunctions.h"
#include "CommonRegexes.h"
#include "GameVersion.h"
#include "Log.h"
#include "OSCompatibilityLayer.h"
#include "ParserHelpers.h"
#include <fstream>
auto laFabricaDeColor = commonItems::Color::Factory();

Configuration::Configuration(const commonItems::ConverterVersion& converterVersion)
{
	Log(LogLevel::Info) << "Reading configuration file";
	registerKeys();
	parseFile(std::filesystem::path("configuration.txt"));
	clearRegisteredKeywords();
	setOutputName();
	verifyCK3Path();
	verifyCK3Version(converterVersion);
	verifyEU4Path();
	verifyEU4Version(converterVersion);
	Log(LogLevel::Progress) << "3 %";
}

Configuration::Configuration(std::istream& theStream)
{
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
	setOutputName();
}

void Configuration::registerKeys()
{
	registerKeyword("SaveGame", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString path(theStream);
		SaveGamePath = path.getString();
		Log(LogLevel::Info) << "Save Game set to: " << SaveGamePath;
	});
	registerKeyword("CK3directory", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString path(theStream);
		CK3Path = path.getString();
	});
	registerKeyword("EU4directory", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString path(theStream);
		EU4Path = path.getString();
	});
	registerKeyword("CK3DocDirectory", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString path(theStream);
		CK3DocPath = path.getString();
	});
	registerKeyword("output_name", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString nameStr(theStream);
		outputName = nameStr.getString();
		Log(LogLevel::Info) << "Output name set to: " << outputName;
	});
	registerKeyword("start_date", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString startDateString(theStream);
		startDate = STARTDATE(std::stoi(startDateString.getString()));
		Log(LogLevel::Info) << "Start date set to: " << startDateString.getString();
	});
	registerKeyword("i_am_hre", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString hreString(theStream);
		iAmHre = I_AM_HRE(std::stoi(hreString.getString()));
		Log(LogLevel::Info) << "HRE set to: " << hreString.getString();
	});
	registerKeyword("dejure", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString dejureString(theStream);
		dejure = DEJURE(std::stoi(dejureString.getString()));
		Log(LogLevel::Info) << "DeJure set to: " << dejureString.getString();
	});
	registerKeyword("split_vassals", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString splitVassalsString(theStream);
		splitVassals = SPLITVASSALS(std::stoi(splitVassalsString.getString()));
		Log(LogLevel::Info) << "Split Vassals set to: " << splitVassalsString.getString();
	});
	registerKeyword("shatter_empires", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString shatterEmpiresString(theStream);
		shatterEmpires = SHATTER_EMPIRES(std::stoi(shatterEmpiresString.getString()));
		Log(LogLevel::Info) << "Shatter Empires set to: " << shatterEmpiresString.getString();
	});
	registerKeyword("shatter_level", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString shatterLevelString(theStream);
		shatterLevel = SHATTER_LEVEL(std::stoi(shatterLevelString.getString()));
		Log(LogLevel::Info) << "Shatter Level set to: " << shatterLevelString.getString();
	});
	registerKeyword("shatter_hre_level", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString shatterHRELevelString(theStream);
		shatterHRELevel = SHATTER_HRE_LEVEL(std::stoi(shatterHRELevelString.getString()));
		Log(LogLevel::Info) << "Shatter HRE Level set to: " << shatterHRELevelString.getString();
	});
	registerKeyword("siberia", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString siberiaString(theStream);
		siberia = SIBERIA(std::stoi(siberiaString.getString()));
		Log(LogLevel::Info) << "Siberia set to: " << siberiaString.getString();
	});
	registerKeyword("sunset", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString sunsetString(theStream);
		sunset = SUNSET(std::stoi(sunsetString.getString()));
		Log(LogLevel::Info) << "Sunset set to: " << sunsetString.getString();
	});
	registerKeyword("dynamicInstitutions", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString dynamicInstitutionsString(theStream);
		dynamicInstitutions = INSTITUTIONS(std::stoi(dynamicInstitutionsString.getString()));
		Log(LogLevel::Info) << "Dynamic Institutions set to: " << dynamicInstitutionsString.getString();
	});
	registerKeyword("development", [this](const std::string& unused, std::istream& theStream) {
		const commonItems::singleString developmentString(theStream);
		development = DEVELOPMENT(std::stoi(developmentString.getString()));
		Log(LogLevel::Info) << "Development set to: " << developmentString.getString();
	});
	registerKeyword("multiprovdevtransfer", [this](std::istream& theStream) {
		const commonItems::singleString multiprovdevtransferString(theStream);
		multiProvinceDevelopmentTransfer = MULTIPROVDEVTRANSFER(std::stoi(multiprovdevtransferString.getString()));
		Log(LogLevel::Info) << "Multiple Province Dev Transfer set to: " << multiprovdevtransferString.getString();
	});
	registerKeyword("dynasticNames", [this](std::istream& theStream) {
		const commonItems::singleString dynasticNamesString(theStream);
		dynasticNames = DYNASTICNAMES(std::stoi(dynasticNamesString.getString()));
		Log(LogLevel::Info) << "Dynastic Names set to: " << dynasticNamesString.getString();
	});
	registerKeyword("discoveredBy", [this](std::istream& theStream) {
		const commonItems::singleString discoveredByString(theStream);
		discoveredBy = DISCOVEREDBY(std::stoi(discoveredByString.getString()));
		Log(LogLevel::Info) << "Discovered By set to: " << discoveredByString.getString();
	});
	registerKeyword("tribalConfederations", [this](std::istream& theStream) {
		const commonItems::singleString tribalConfederationsString(theStream);
		tribalConfederations = TRIBALCONFEDERATIONS(std::stoi(tribalConfederationsString.getString()));
		Log(LogLevel::Info) << "Tribal Confederations set to: " << tribalConfederationsString.getString();
	});
	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
}

void Configuration::verifyCK3Path()
{
	if (!commonItems::DoesFolderExist(CK3Path))
		throw std::runtime_error(CK3Path.string() + " does not exist!");
	// TODO: OSX and Linux paths are speculative
	if (!commonItems::DoesFileExist(CK3Path / "binaries/ck3.exe") && !commonItems::DoesFileExist(CK3Path / "CK3game") &&
		 !commonItems::DoesFileExist(CK3Path / "binaries/ck3"))
		throw std::runtime_error(CK3Path.string() + " does not contain Crusader Kings 3!");
	if (!commonItems::DoesFileExist(CK3Path / "game/map_data/positions.txt"))
		throw std::runtime_error(CK3Path.string() + " does not appear to be a valid CK3 install!");
	Log(LogLevel::Info) << "\tCK3 install path is " << CK3Path;
	CK3Path += "/game/"; // We're adding "/game/" since all we ever need from now on is in that subdirectory.
}

void Configuration::verifyEU4Path() const
{
	if (!commonItems::DoesFolderExist(EU4Path))
		throw std::runtime_error(EU4Path.string() + " does not exist!");
	if (!commonItems::DoesFileExist(EU4Path / "eu4.exe") && !commonItems::DoesFileExist(EU4Path / "eu4"))
		throw std::runtime_error(EU4Path.string() + " does not contain Europa Universalis 4!");
	if (!commonItems::DoesFileExist(EU4Path / "map/positions.txt"))
		throw std::runtime_error(EU4Path.string() + " does not appear to be a valid EU4 install!");
	Log(LogLevel::Info) << "\tEU4 install path is " << EU4Path;
}

void Configuration::setOutputName()
{
	if (outputName.empty())
	{
		outputName = SaveGamePath.filename();
	}
	auto outputNameString = outputName.stem().string();
	outputNameString = replaceCharacter(outputNameString, '-');
	outputNameString = replaceCharacter(outputNameString, ' ');

	outputName = commonItems::normalizeUTF8Path(outputNameString);
	Log(LogLevel::Info) << "Using output name " << outputName.string();
}

void Configuration::verifyCK3Version(const commonItems::ConverterVersion& converterVersion) const
{
	const auto CK3Version = GameVersion::extractVersionFromLauncher(CK3Path / "../launcher/launcher-settings.json");
	if (!CK3Version)
	{
		Log(LogLevel::Error) << "CK3 version could not be determined, proceeding blind!";
		return;
	}

	Log(LogLevel::Info) << "CK3 version: " << CK3Version->toShortString();

	if (converterVersion.getMinSource() > *CK3Version)
	{
		Log(LogLevel::Error) << "CK3 version is v" << CK3Version->toShortString() << ", converter requires minimum v"
									<< converterVersion.getMinSource().toShortString() << "!";
		throw std::runtime_error("Converter vs CK3 installation mismatch!");
	}
	if (!converterVersion.getMaxSource().isLargerishThan(*CK3Version))
	{
		Log(LogLevel::Error) << "CK3 version is v" << CK3Version->toShortString() << ", converter requires maximum v"
									<< converterVersion.getMaxSource().toShortString() << "!";
		throw std::runtime_error("Converter vs CK3 installation mismatch!");
	}
}

void Configuration::verifyEU4Version(const commonItems::ConverterVersion& converterVersion) const
{
	const auto EU4Version = GameVersion::extractVersionFromLauncher(EU4Path / "launcher-settings.json");
	if (!EU4Version)
	{
		Log(LogLevel::Error) << "EU4 version could not be determined, proceeding blind!";
		return;
	}

	Log(LogLevel::Info) << "EU4 version: " << EU4Version->toShortString();

	if (converterVersion.getMinTarget() > *EU4Version)
	{
		Log(LogLevel::Error) << "EU4 version is v" << EU4Version->toShortString() << ", converter requires minimum v"
									<< converterVersion.getMinTarget().toShortString() << "!";
		throw std::runtime_error("Converter vs EU4 installation mismatch!");
	}
	if (!converterVersion.getMaxTarget().isLargerishThan(*EU4Version))
	{
		Log(LogLevel::Error) << "EU4 version is v" << EU4Version->toShortString() << ", converter requires maximum v"
									<< converterVersion.getMaxTarget().toShortString() << "!";
		throw std::runtime_error("Converter vs EU4 installation mismatch!");
	}
}
